### ISSUE TITLE
App Rendering Protocol (ARP)

### ISSUE THREAD

**Issue by lite20** [NodeOS/NodeOS#215]
So after closing the original thread where the idea was designed, we know are that much closer to our graphic systems. With remote renderers complete and local one's in the works, one question still remains.

"So how do I make a window?"

Whether we treat the content of a window as a page, a canvas, or whatnot, the actual process to create this, and interface to work with it needs to be standard for apps to work across NodeOS platforms. Ideally, a well designed platform can also have a shell to allow it to run on platforms other than NodeOS.

The simplest way to go about this is to create a module. This modules sole job would be to tell whatever window manager happens to be on the device, to create a window (via some interface). And allow the app to render to the window. That is all, as any additional manager specific features could be included separately in different modules by that window manager's maintenance team.

Seeing that apps need to work in the browser, this 'module' could be a dual that work both as an import or a nodejs module. I will create a repository for us (the nodeos community) to begin hacking away at a system that works.

**Comment by lite20** [comment]
> needs to be standard for apps to work across NodeOS platforms.

I meant to say, it needs to be standardized for it to work across NodeOS window managers. As not all managers may chose to run apps fullscreen, they need to be asked first and interacted with when an app works with a window.

**Comment by Coretool** [comment]
I think we should make the apps working no matter what window manager (I'll use WD because it's faster) with something like a manifest.json.

**Comment by Coretool** [comment]
And don't forget that apps which do not run on a Desktop will always be rendered as full screen apps, since the framebuffer size will be set to full screen...

**Comment by piranna** [comment]
Isn't this fixed with the URLs design I proposed to launch GUI apps backend
servers? Or are you talking about a common protocol for the desktop
environments to interact with the window manager (maximize, minimize,
close...)? If so, the important part is to have a common API that all apps
could agree, and for this se could write a library that overwrite the
current methods for the browser Window object (or implement them for
non-browser environments). Is it really needed to store some state on the
Desktop Environment backend server? I think the Window Manager could (and
should) be implemented purely client side as Os.js does...

**Comment by piranna** [comment]
In other words: KISS (Keep It Simple, Stupid) and DRY (Don't Repeat
Yourself) rules ;-)

**Comment by lite20** [comment]
> And don't forget that apps which do not run on a Desktop will always be rendered as full screen apps,

Yes because of this the "window size" params would only be preferences.

> Or are you talking about a common protocol for the desktop
environments to interact with the window manager (maximize, minimize,
close...)?

I would assume the window manager would be a part of the desktop environment and that the developer of the desktop environment should ensure the two are compatible. By a rendering protocol I mean, say I'm an app. How do I make a window to show my content to? Apps ought not to be limited to one frame. The ARP would specify how apps ough to simply create window. From there on, the interactions with it and such are all specific to the window manager.

**Comment by piranna** [comment]
> Yes because of this the "window size" params would only be preferences.

I'm not sure if there's already support for this on HTML/Javascript or only as a parameter when calling [window.open()](https://developer.mozilla.org/en-US/docs/Web/API/Window/open)... If so we should use it instead of invent our own mechanism.


> I would assume the window manager would be a part of the desktop environment and that the developer of the desktop environment should ensure the two are compatible.

I agree.


> By a rendering protocol I mean, say I'm an app. How do I make a window to show my content to?

Simple: use `window.open()` :-)

* Running full-screen without a desktop manager? The browser native method will be used.
* Running inside a HTML5 Desktop Environment? The Desktop Environment would have overwrite the `window.open()` method of the app window `iframe` so it will call the Desktop Environment internal mechanism to create a new window (mostly create a new iframe, but it would be whatever).
* Running on an "alien" environment (for example a framebuffer-based renderer)? That environment should emulate the `window.open()` method someway.


> Apps ought not to be limited to one frame. The ARP would specify how apps ough to simply create window. From there on, the interactions with it and such are all specific to the window manager.

I agree, but as I've said, I think it's just enough that all Desktop Environments and Window Managers use a common API, and the most obvious one is to use the methods from the browser [Window](https://developer.mozilla.org/en-US/docs/Web/API/Window) object.

**Comment by zodern** [comment]
In Silk I am working on a new app framework which is compatible with [electron](http://electron.atom.io/). [Stretto](https://github.com/benkaiser/stretto) works with the new framework.

It requires nodeIntegration to be disabled in all windows and will not support preloading files. I have only found a couple electron apps that meets those requirements. But all apps made for Silk will run on Windows, OSX and Linux with no modifications.

**Comment by piranna** [comment]
`nodeIntegration` is what makes possible to access Node.js APIs directly from the HTML/client side code, isn't it? +1 to require it to be disabled, makes it more difficult to program but the architecture is more elegant and flexible.

**Comment by lite20** [comment]
> Simple: use window.open() :-)

By using window.open we actually break compatibility with existing browsers. Since window.open() exists already, when called in a browser it would create an additional window even though the code is already being run in one where as on node it doesn't have one. We also can't immediately create a window when imported into the module (required) if it's a module that is, because the app may not want to use the window right away (and hence the developers might have preferred to create it later)

**Comment by lite20** [comment]
The proposal solution is to implement both functions. One function would be used only at start to create the initial window browsers would already have. If called in a browser this would throw an error. However, developers can use a simple pollyfill to solve. If the function name was buildFrame then
`if(window.buildFrame) {
    window.buildFrame("window", 720, 480);
}`

quick and easy fix!

**Comment by lite20** [comment]
> I am working on a new app framework fork Silk which is compatible with electron. Stretto already runs in Silk.

If you chose to support nodeOS, you could easily just import said 'window' module and expand it to your users. I'm reserving the npm name 'frame' for it.

**Comment by piranna** [comment]
> By using window.open we actually break compatibility with existing browsers. Since window.open() exists already, when called in a browser it would create an additional window even though the code is already being run in one where as on node it doesn't have one.

I understand your concerns, but you are taking a bad point of view: we are writting a Windows Manager using HTML `iframes` as containers to host the GUI apps "windows" content. The point is, you are missing that fact that each one of these `iframe`s has **its own** `window` object, and independent from the browser "main frame" one (the Desktop Environment / WIndow Manager). Taking this in account, nothing forbids us to inject code that overwrite the `window.open()` method inside that `iframe` to do a `postMessage()` call to the parent frame (the Window manager) to create a new `iframe` for the new GUI app "window" instead of having the browser default behaviour of create a new browser window, and in fact this is something similar to what Firefox does to show modal windows inside the browser rendering region instead of OS level independent windows. The main problem would be how to know when to inject that code on the iframe, since it must be on each `iframe` navigation change after the current page clean-up (if any) and just before the new page start downloading and executing new content.

**Comment by piranna** [comment]
> Taking this in account, nothing forbids us to inject code that overwrite the window.open() method inside that iframe to do a postMessage() call to the parent frame (the Window manager) to create a new iframe for the new GUI app "window" instead of having the browser default behaviour of create a new browser window

Well... here we have something that could be discussed, what protocol to follow in this `postMessage()` based communication, this way the injected code to overwrite the behaviour of the `window` object on the `iframes` would be common and the diferent Windows Manager only would need to listen for this messages and process them :-) The most simple and minimal solution here is to use a 1:1 mapping from the `window` object method names and their calls positional arguments to JSON-RPC requests. Only problem I would find here is that I believe most of these methods are sincronous... :-/

**Comment by lite20** [comment]
> Well... here we have something that could be discussed, what protocol to follow in this postMessage() based communication, this way the injected code to overwrite the behaviour of the window object on the iframes would be common and the diferent Windows Manager only would need to listen for this messages and process them :-) The most simple and minimal solution here is to use a 1:1 mapping from the window object method names and their calls positional arguments to JSON-RPC requests. Only problem I would find here is that I believe most of these methods are sincronous... :-/

Ah. I haven't worked much with JSON-RPC so I wouldn't be to useful here.

> I understand your concerns, but you are taking a bad point of view: we are writting a Windows Manager using HTML iframes as containers to host the GUI apps "windows" content. The point is, you are missing that fact that each one of these iframes has its own window object, and independent from the browser "main frame" one (the Desktop Environment / WIndow Manager).

Ah okay, that makes a lot more sense

**Comment by piranna** [comment]
> Ah okay, that makes a lot more sense

I'm happy you now got it ;-)

**Comment by lite20** [comment]
Okay so in that case, it's the job of the window manager to inject the code into the iframe. No module needed. Additionally, since we only implement a different way for window.open to work, and the injected code speaks back to it's window manager which was the platform that injected it in the first place, the interface between the inject and the manager also don't need to be standardized.

So only thing we needed to discuss is that new windows are created with `window.open()`. Excellent. I've been working on a desktop environment by the way. If you wanted to check it out and try it in NodeOS, let me know!

**Comment by piranna** [comment]
> Okay so in that case, it's the job of the window manager to inject the code into the iframe. No module needed. Additionally, since we only implement a different way for window.open to work, and the injected code speaks back to it's window manager which was the platform that injected it in the first place, the interface between the inject and the manager also don't need to be standardized.

Yes, that's it :-) Problem seems to be `iframes` has a lot of limitations on browsers mainly for security issues (for example is not possible to access the content of `iframes` of a diferent domain, nor some pages like Google or Facebook can be viewed from inside `iframes`) but also seems there's no way on Javascript to know when an `iframe` is going to change a page until it has fully loaded... :-( Not sure if FirefoxOS `browser` tag would be useful, but seems it will be difficult anyway to be able to use a general and transparent solution... :-/


> So only thing we needed to discuss is that new windows are created with window.open(). Excellent. I've been working on a desktop environment by the way. If you wanted to check it out and try it in NodeOS, let me know!

You don't need to ask, just share the link :-)

**Comment by lite20** [comment]
> Yes, that's it :-) Problem seems to be iframes has a lot of limitations on browsers mainly for security issues (for example is not possible to access the content of iframes of a diferent domain, nor some pages like Google or Facebook can be viewed from inside iframes) but also seems there's no way on Javascript to know when an iframe is going to change a page until it has fully loaded... :-( Not sure if FirefoxOS browser tag would be useful, but seems it will be difficult anyway to be able to use a general and transparent solution... :-/

I'm not sure if you were aware but I've been working on a web browser. My cousin has contributed a bit, and @Coretool is in on the project. Should we run in with too many issues, eventually it may serve to be a nice alternative to content_frame or whatever other systems we attempt to implement as we'd have a complete understanding of the under working system.

Iframes do have issues. I'm experimenting to see if something similar to an iframe can be programatically created in JS. Then we could use said modded iframe in place of the actual to try and escape some of the issues with it.

So Google and Facebook wouldn't be visible if we use iframes. That may be a major issue. What other options do we have aside from iframes?

**Comment by lite20** [comment]
> You don't need to ask, just share the link :-)

It's called Ryutai and I just threw it on NPM to experiment on NodeOS. It's dependencies currently surpass what NodeOS's default ram provides but I believe I have un-needed dependencies and so I'm looking through to clean them up so it'll work.

**Comment by Coretool** [comment]
@lite20 You may want to have a look at electron (aka atom-shell) They have got a 'web view' tag. It's basically an iframe with some extras... (I don't have a link, but it's on GitHub...)

**Comment by lite20** [comment]
@Coretool I know electron :P I didn't know it had said web view though. I'll look into it!

**Comment by piranna** [comment]
> I'm not sure if you were aware but I've been working on a web browser.

Problem is not to work on our own web browser where we could do whatever we wants, point is to be able to use regular web browsers on normal PCs or mobile phones as terminals, that's the tricky part and the reason why I was encouraging so much to follow standard technologies and make the GUI based on HTML5...


> @lite20 You may want to have a look at electron (aka atom-shell) They have got a 'web view' tag. It's basically an iframe with some extras... (I don't have a link, but it's on GitHub...) Maybe we can borrow some code ;)

It could be an option similar to the `browser` tag on FirefoxOS, but I'm not sure if Electron can be (easily) compiled statically nor using framebuffer/EGL... And of course, there's still the compatiblity problem with regular browsers on regulars PCs...

**Comment by lite20** [comment]
@piranna Velocity, the name of the browser, will be released on all platforms (excluding IOS on debut) so at some point it will be a "regular web browser" for "normal PCs" :D.

The goal was simply to create a browser implementing new features as they came. Being easy to understand and modify anyone could embed it in whatever their projects may be

**Comment by piranna** [comment]
By "regular web browser" I means that users don't need to install anything to connect to a NodeOS server remotely and start a GUI session. Anyway, your project seems insteresting... :-)

**Comment by lite20** [comment]
@piranna ahh okay.

https://bitbucket.org/Lite20/velocity

Here's the project. It's just a baby but there it is :)
More progress is on the beta branch. Right now I've implemented the shell in Java for easy portability but I may switch to c/c++ at some point experimentally.

**Comment by lite20** [comment]
most of the browsers runtime occurs inside V8. The browser itself is coded in js, however the Java shell extends v8's native features adding more functionality for the browser, and inturn webpages to harness. (also lets the js create a window and render to it)

**Comment by lite20** [comment]
> More progress is on the beta branch.

Correction, alpha branch. Alpha is however untested

**Comment by Coretool** [comment]
@piranna I've looked into building electron and I don't think that it would be easy to make it render to the linuxfb...

**Comment by piranna** [comment]
> @piranna I've looked into building electron and I don't think that it would be easy to make it render to the linuxfb...

Sh*t... :-(

**Comment by lite20** [comment]
@Coretool Yeah... no. XD have fun with that. It's modified CEF impements x11 on linux (maybe support wayland) but the struggle it would take to implement it for the fb is real.

> Sh*t... :-(

What other options do we have?

**Comment by Coretool** [comment]
Just stick with noGui (or start with it ^^) and iframes ^^

**Comment by piranna** [comment]
Good idea, let's get something working and later we'll move on.

**Comment by lite20** [comment]
@piranna @Coretool sounds good. Awaiting a working noGui build. Let me know if you guys get anywhere!

**Comment by junland** [comment]
Congrats! Hope to see more down the pipeline! :)

**Comment by piranna** [comment]
> @piranna @Coretool sounds good. Awaiting a working noGui build. Let me
know if you guys get anywhere!

Could someone take a look on the docs how to give an URL to QtWebkit, at
least a hardcoded one? I only get a blank page (the same as @formula1 with
its experiment with webkit.js...).

**Comment by Coretool** [comment]
@piranna Take a look at noGui. (main.cc) It does that ^^

**Comment by piranna** [comment]
> Take a look at noGui. (main.cc) It does that ^^

Lol :-P Yes, you're right. We need more error checking there, since at this moment it hides the console output and gets control of the keyboard and only way is to swith off the machine, that's tedious to test...

**Comment by Coretool** [comment]
Yeah... I've to add the keyboard shortcuts and so on.

**Comment by formula1** [comment]
http://doc.qt.io/qt-5/qwebview.html#load

Is this the method you're looking for?

I'm going to checkout breifly how embeddable servo is. They recently (I think last week) showed an intense proof of concept that rendered Rediculous things at > 60 fps. They acheived this through offloading most of the work to the GPU (so GPU-less machines may still be in trouble) but they are really doing great stuff over there. I really wouldn't be too suprised if a rust javascript engine becomes standard over time.

**Comment by Coretool** [comment]
@formula1 Yes, it's this method.

**Comment by piranna** [comment]
> I'm going to checkout breifly how embeddable servo is. They recently (I
think last week) showed an intense proof of concept that rendered
Rediculous things at > 60 fps. They acheived this through offloading most
of the work to the GPU (so GPU-less machines may still be in trouble) but
they are really doing great stuff over there. I really wouldn't be too
suprised if a rust javascript engine becomes standard over time.

It would be good to have several rendering engines to compare. I'm going to
try to find some time to fix the ICU static compilation, but if someone
wants to take a look on it or avout getting noGUI to render something, any
help will be welcome. In that way, I was thinking we could be able to
enable X11 support to make it easier to test and when we get something
usable disable it again, what do you think?

**Comment by lite20** [comment]
> It would be good to have several rendering engines to compare.

I'm now contending with the struggle of designing and constructing my own mostly taking inspiration from Blink. Hopefully I'll be able to give more insight on possible solutions with my greater understanding of web engines when I'm done this project (Velocity)

**Comment by Coretool** [comment]
@piranna I think I have to add some logging to noGui, it doesn't throw errors at all :/ ... I know that the 'normal' desktop version works but I've really no plan, why the static one doesn't... Enabling X11 would be great for testing and maybe a first release since it would make the qt stuff far easier (querying the display size, setting the view size etc ... even if I'm not sure if this would still be needed )
